Overwinter transaction format

bitcartel opened the pull request:
Implements ZIP 202.

Closes #2906.

str4d commented on src/main.cpp:
```diff
@@ -872,6 +875,18 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+    // Check Overwinter flag has been set if transaction version 3 or later
+    if (tx.nVersion >= 3 && !tx.fOverwinter) {
+        return state.DoS(100, error("CheckTransaction(): version requires overwinter flag to be set"),
+                         REJECT_INVALID, "overwinter-flag-not-set");
```
I'm not sure whether this is necessary; this should be a parser error. It probably doesn't *hurt*, though that's not really something I want to say about consensus rules...

str4d replied:
Hmm, it occurs to me that we should use the masked version in the signature hash, which means this boolean will get ORed in, which means we probably do want some kind of check (to defend against internal mistakes, not bad transactions from peers - those should be caught as parser errors).

I would probably implement the above in some `GetHeader()` function, rather than doing it inside the sighash code, because then we know it's being masked the same way everywhere.

str4d commented on src/main.cpp:
```diff
@@ -872,6 +875,18 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+    // Check branch id has set for Overwinter transactions
+    if (tx.nVersion >= 3 && tx.nFormatBranchId==0) {
+        return state.DoS(100, error("CheckTransaction(): overwinter requires branch id to be set"),
+                         REJECT_INVALID, "overwinter-branch-id-not-set");
```
This OTOH is probably a good check.

str4d commented on src/primitives/transaction.cpp:
```diff
@@ -248,12 +254,23 @@ unsigned int CTransaction::CalculateModifiedSize(unsigned int nTxSize) const
-        vin.size(),
-        vout.size(),
-        nLockTime);
+    if (nVersion == 1 || nVersion == 2) {
```
`if (nVersion == 1 || nVersion == 2 || !fOverwintered) {` per https://github.com/zcash/zips/pull/133#issuecomment-361616008

str4d commented on src/primitives/transaction.cpp:
```diff
@@ -248,12 +254,23 @@ unsigned int CTransaction::CalculateModifiedSize(unsigned int nTxSize) const
+            vin.size(),
+            vout.size(),
+            nLockTime);
+    } else if (nVersion == 3) {
```
`} else if (fOverwintered && nVersion == 3) {` per https://github.com/zcash/zips/pull/133#issuecomment-361616008

bitcartel replied:
This and the above comment are now simplified to:
```
if (!fOverwintered) { ... } else if (nVersion == 3) { ... }
```

str4d commented on src/primitives/transaction.h:
```diff
@@ -351,8 +357,28 @@ class CTransaction
+            *const_cast<bool*>(&fOverwinter) = header >> 31;
+            *const_cast<int32_t*>(&this->nVersion) = header & 0x7FFFFFFF;
+        } else {
+            if (nVersion == 3) {
```
`if (fOverwintered) {`

str4d commented on src/primitives/transaction.h:
```diff
@@ -20,6 +20,8 @@
 #include "zcash/JoinSplit.hpp"
 #include "zcash/Proof.hpp"
 
+#include <iostream>
```
Left over from debugging?

bitcartel replied:
Ha.. you got me... but this is a WIP :-)

str4d commented on src/main.cpp:
```diff
@@ -843,6 +844,8 @@ unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& in
 bool CheckTransaction(const CTransaction& tx, CValidationState &state,
                       libzcash::ProofVerifier& verifier)
 {
+    // FIXME: Check transaction with context - check against contect of local accepted branch here?
```
The target branch ID is probably sufficient context; see my later comment.

bitcartel replied:
I don't understand this comment.

str4d commented on src/main.cpp:
```diff
@@ -1059,6 +1074,8 @@ CAmount GetMinRelayFee(const CTransaction& tx, unsigned int nBytes, bool fAllowF
 bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransaction &tx, bool fLimitFree,
                         bool* pfMissingInputs, bool fRejectAbsurdFee)
 {
+    // FIXME: Overwinter txs check expiry height here?
```
I guess we add the field in this PR, and then @arcalinea rebases her PR on top with the logic?

bitcartel replied:
Yes, that could work, since dependency is: activation logic <-- tx format <-- tx expity logic

bitcartel replied:
Added

str4d commented on src/primitives/transaction.cpp:
```diff
@@ -145,9 +145,10 @@ std::string CTxOut::ToString() const
-
-CMutableTransaction::CMutableTransaction() : nVersion(CTransaction::MIN_CURRENT_VERSION), nLockTime(0) {}
-CMutableTransaction::CMutableTransaction(const CTransaction& tx) : nVersion(tx.nVersion), vin(tx.vin), vout(tx.vout), nLockTime(tx.nLockTime),
+// FIXME: we need a default value for branch id and expiry height, or leave it as 0?
```
Default branch ID should match the default version, so if we default to v1 then use 0. I'm not quite sure what we do about the transition; probably just leave it defaulting to the Sprout version until after the upgrade is live (so manually setting Overwinter versions everywhere), and then we can change it once Overwinter has activated (and clean up the usage everywhere).

Default expiry height should be somewhere in #2874.

bitcartel replied:
Agreed.  I think for now we leave the defaults to v1.

str4d commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -391,3 +391,107 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+    CMutableTransaction mtx;
+    mtx.fOverwinter = true;
+    mtx.nVersion = 3;
+    mtx.nFormatBranchId =0xDEADBEEF;
```
Given that the serialization will check this value:

```
mtx.nFormatBranchId = NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId;
```

bitcartel replied:
I rebased.  I'm keeping 0xDEADBEEF for now since UPGRADE_OVERWINTER doesn't exist right now.  Marked as FIXME in the code.

str4d commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -391,3 +391,107 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+    mtx.vjoinsplit.resize(0);
+    mtx.fOverwinter = true;
+    mtx.nVersion = 3;
+    mtx.nFormatBranchId =0xDEADBEEF;
```
Ditto.

bitcartel commented on src/primitives/transaction.h:
```diff
@@ -332,6 +335,9 @@ class CTransaction
     const int32_t nVersion;
+    const bool fOverwinter;
+    const uint32_t nFormatBranchId;
+    const uint32_t nExpiryHeight; // FIXME:maybe int32_t since block height typically int?
```
Hmm, see my comment here https://github.com/zcash/zips/pull/131#discussion_r164284115 about UINT_MAX -> INT

bitcartel replied:
This was resolved in disussion, see: https://github.com/zcash/zips/pull/131#issuecomment-363502380

bitcartel commented on src/primitives/transaction.h:
```diff
@@ -425,8 +454,28 @@ struct CMutableTransaction
+                READWRITE(*const_cast<uint32_t*>(&header));
+            } else {
+                // When serializing v1 and v2, the 4 byte header is nVersion
+                READWRITE(*const_cast<int32_t*>(&this->nVersion));
```
Dreaded copy and paste...

bitcartel commented on src/primitives/transaction.h:
```diff
@@ -351,8 +357,28 @@ class CTransaction
+        if (ser_action.ForRead()) {
+            // When deserializing, unpack the 4 byte header to extract fOverwinter and nVersion.
+            uint32_t header;
+            READWRITE(*const_cast<uint32_t*>(&header));
```
Copy and paste..

bitcartel commented on src/primitives/transaction.h:
```diff
@@ -351,8 +357,28 @@ class CTransaction
+            }
+        }
         nVersion = this->nVersion;
+        bool isOverwinterV3 = (fOverwinter && nVersion==3);
```
@daira I think that `nFormatBranchId` might be better renamed `nFormatIntroducedBranchId` to eliminate any confusion!

bitcartel replied:
Resolved, will update PR.

bitcartel commented:
Pushed update to address comments so far.

str4d commented on src/main.cpp:
```diff
@@ -1074,8 +1111,35 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
+
+    uint32_t consensusBranchId = 0;
+    // FIXME: get the branch id given current block height... for now set to 0xDEADBEEF
+    if (tx.fOverwintered && tx.nVersion >= 3) {
```
We must not set `consensusBranchId` based on transaction information, as that creates a circular consensus rule dependency. Use the branch ID that will be associated with the next block that the transaction could be mined in:

```
auto consensusBranchId = CurrentEpochBranchId(chainActive.Height() + 1, Params().GetConsensus());
```

bitcartel replied:
Removed conditional.  Added check to enforce `fOverwintered` must be true if Overwinter is active.

str4d commented on src/main.cpp:
```diff
@@ -1074,8 +1111,35 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
+    uint32_t consensusBranchId = 0;
+    // FIXME: get the branch id given current block height... for now set to 0xDEADBEEF
+    if (tx.fOverwintered && tx.nVersion >= 3) {
+        // FIXME: should we validate against tx.nFormatBranchID here?
```
No; `tx.nFormatBranchId` is only for parsing.

str4d commented on src/main.h:
```diff
@@ -336,13 +336,13 @@ bool ContextualCheckInputs(const CTransaction& tx, CValidationState &state, cons
  * @return True if all outputs (scriptPubKeys) use only standard transaction forms
  */
-bool IsStandardTx(const CTransaction& tx, std::string& reason);
+bool IsStandardTx(const CTransaction& tx, std::string& reason, uint32_t formatBranchId = 0);
```
`consensusBranchId`

str4d commented on src/primitives/transaction.h:
```diff
@@ -351,11 +356,36 @@ class CTransaction
+            if (fOverwintered) {
+                // When serializing v3, assemble the 4 byte header from fOverwintered and nVersion
+                uint32_t header = (fOverwintered << 31) | this->nVersion;
+                READWRITE(*const_cast<uint32_t*>(&header));
```
No need for `const_cast` here.

str4d commented on src/primitives/transaction.h:
```diff
@@ -351,11 +356,36 @@ class CTransaction
+        } else {
+            if (fOverwintered) {
+                // When serializing v3, assemble the 4 byte header from fOverwintered and nVersion
+                uint32_t header = (fOverwintered << 31) | this->nVersion;
```
I think we can simplify the code and just use this in both cases, since `fOverwintered << 31` will be zero if `fOverwintered == false`. @daira?

bitcartel replied:
It could be 1 << 31 or 0x80000000

str4d replied:
Right; what I mean is that if `fOverwintered == false`, then `header == this->nVersion`, so we can get rid of the `if (fOverwintered)` and the other branch.

bitcartel replied:
Simplified.

str4d commented on src/primitives/transaction.h:
```diff
@@ -351,11 +356,36 @@ class CTransaction
+        if (isOverwinterV3) {
+            READWRITE(*const_cast<uint32_t*>(&nExpiryHeight));
+        }
         if (nVersion >= 2) {
```
Do we want to restrict this as well (wrt not having soft-forking capacity for Overwinter)? ie:
```
if (isOverwinterV3 || (!fOverwintered && nVersion >= 2)) {
```

str4d replied:
It might be easier to do this simply by throwing an exception earlier if `fOverwintered && nVersion > 3`.

bitcartel replied:
I'm going to remove the exception throwing as a test showed that it would resulted in unexpected code path.  When assigning via `=` GetHash would be invoked which involves serailiazing.

bitcartel replied:
Ok I'm not going to remove this anymore based on your later comment.

daira replied:
I'm dubious about the exception; are exceptions typically used to signal errors in deserialization?

str4d replied:
They definitely are in the Bitcoin-inherited (de)serialization code:

- https://github.com/zcash/zcash/blob/master/src/serialize.h#L286-L305
- https://github.com/zcash/zcash/blob/master/src/serialize.h#L455-L457
- https://github.com/zcash/zcash/blob/master/src/serialize.h#L766-L768
  - We added this one.
- https://github.com/zcash/zcash/blob/master/src/main.cpp#L5411-L5417
  - Catches parsing exceptions in network messages.
- https://github.com/zcash/zcash/blob/master/src/core_read.cpp#L100-L105
  - Catches parsing exceptions in hex transactions from RPC method parameters.

I'm of the opinion that an exception is better than having a `C[Mut]Transaction` with undefined state.

str4d commented on src/main.cpp:
```diff
@@ -1074,8 +1111,35 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
+    if (tx.fOverwintered &&
+        tx.nVersion == 3 &&
+        tx.nFormatBranchId != 0 &&
+        tx.nExpiryHeight != std::numeric_limits<uint32_t>::max())
```
This should go after `CheckTransaction()`; I'd put it between the coinbase check and `IsStandardTx()` check. Then we can remove the `tx.nFormatBranchId != 0` conditional.

str4d commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -391,3 +391,115 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+    mtx.nVersion = 3;
+    mtx.nFormatBranchId =0xDEADBEEF;
+    // FIXME: Eventually replace with
+    // NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId;
```
(Posting again elsewhere because the original one shows up as outdated.)

If you rebase on current #2898, this parameter exists 😃 (https://github.com/zcash/zcash/pull/2898/commits/1eac915de98a5cfec9d4db88bdb0eaea91b283ab)

arcalinea commented on src/gtest/test_mempool.cpp:
```diff
@@ -136,3 +136,66 @@ TEST(Mempool, TxInputLimit) {
+    // NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId;
+
+    // There is no active chain so expiry height is being compared to chainActive.Height() which is -1
+    // To simulate an expired tx, we must set an expiry height less than -1, that is, set a uint32 value
```
This is confusing, there should never be a negative chain height. Can't we compare to a positive integer, or simulate a valid chain height? 

Also, why would an expired tx have a smaller nExpiryHeight than chainheight of `-1`? Should expire if it's >= to the chainheight.

bitcartel replied:
The issue is that when the unit test runs, there's no actual chain when calling into AcceptToMemoryPool, so chainActive.Height() returns -1.  I think a QA test will be better for testing expired txs as it will involve actually relaying txs from one node to another.  Setting tx expiry to -2 is a hack for the test, since it's less than -1, it will not be expired.  However, since the type at the moment is uint32_t and we need it to be int for the comparison against -1, we pick the uint32_t value which is the equivalent of -2 when it gets cast to an int.

arcalinea replied:
Ah ok, seems hacky, but I understand why this is necessary for the unit test now. Thanks for explaining.

arcalinea commented on src/gtest/test_mempool.cpp:
```diff
@@ -136,3 +136,66 @@ TEST(Mempool, TxInputLimit) {
+    EXPECT_FALSE(AcceptToMemoryPool(pool, state2, tx2, false, &missingInputs));
+    EXPECT_EQ(state2.GetRejectReason(), "bad-txns-vin-empty");
+
+    // Check a transaction has not expired, and catch the next error about missing inputs.  Set an expiry height > -1.
```
999 is >= -1, so this should be an expired transaction...?

bitcartel replied:
Do you have a python QA test in your PR?  It would be better to rely on those for actually testing that expiry is working as intended.

arcalinea replied:
Haven't finished, will try to push it up today.

bitcartel commented:
Pushed another update.

- AcceptToMemoryPool, CheckBlock and CheckTransaction* deserve scrutiny.
- Addressed most comments, did not update the expiry height datatype issue.  
- Throwing an exception in transaction.h, has unintended consequence, see FIXME and in test_mempool.  I don't think we need it there but...
- Soft fork >= 3 vs == 3 - need some guidelines here for consistency across code as to what we allow and what the error condition should be.
- Testing expiry height; can we fake a chain to get consensus branch id to match tx format?  Or handle this via qa test in regtest mode?

bitcartel commented:
Just a reminder note to update mininode.py to handle Overwinter tx format and that changes will happen both here and in PR #2919 (for version handshake).  The following tests reference CTransaction.

- bip65-cltv-p2p
- bipdersig-p2p

str4d commented on src/main.cpp:
```diff
@@ -3069,9 +3119,18 @@ bool CheckBlock(const CBlock& block, CValidationState& state,
+    uint32_t consensusBranchId = 0;
+    int futureBlockHeight = chainActive.Height() + 1;
+    const Consensus::Params& params = Params().GetConsensus();
+    if (UPGRADE_ACTIVE == NetworkUpgradeState(futureBlockHeight, params, Consensus::UPGRADE_OVERWINTER)) {
```
Can be simplified to `NetworkUpgradeActive(futureBlockHeight, params, Consensus::UPGRADE_OVERWINTER)`.

Can also be removed entirely; `CurrentEpochBranchId` already contains this logic internally (by virtue of Overwinter being the first upgrade), and we want to always use the branch ID of the current epoch as the consensus branch ID.

str4d commented on src/main.cpp:
```diff
@@ -3069,9 +3119,18 @@ bool CheckBlock(const CBlock& block, CValidationState& state,
+    // FIXME: Need test for CheckBlock 
+    // Get branch id based on height of block which should advance tip
+    uint32_t consensusBranchId = 0;
+    int futureBlockHeight = chainActive.Height() + 1;
```
This will not be correct for upgrades after Overwinter - it will result in Overwinter blocks being checked against the chain tip's consensus rules. We should either pass the block's block height into `CheckBlock()`, or (more likely) pass `consensusBranchId` in.

str4d commented on src/primitives/transaction.h:
```diff
@@ -351,11 +357,39 @@ class CTransaction
+            READWRITE(*const_cast<uint32_t*>(&this->nFormatBranchId));
+        }
+        // FIXME: Do we want to do this?  Will it lead to unexpected code paths?
+        if (fOverwintered && nVersion != 3) {
```
Check `nFormatBranchId` here (probably using the same `tx.nVersion -> tx.nFormatBranchId` mapping as in the consensus rules).

bitcartel replied:
Updated to check version group id too.

str4d commented on src/primitives/transaction.h:
```diff
@@ -351,11 +357,39 @@ class CTransaction
+        }
+        bool isOverwinterV3 = fOverwintered &&
+                              nVersion == 3 &&
+                              nFormatBranchId == NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId;
```
Move this `nFormatBranchId` into its own check (probably using the same `tx.nVersion -> tx.nFormatBranchId` mapping as in the consensus rules) immediately after its `READWRITE` call, and throw an exception if the mapping is invalid.

bitcartel replied:
Going to leave as is for now, can't share code as CMutableTransaction is a struct and not a subclass of CTransaction.

bitcartel commented on src/main.cpp:
```diff
@@ -1074,8 +1111,35 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
+        tx.nFormatBranchId != 0 &&
+        tx.nExpiryHeight != std::numeric_limits<uint32_t>::max())
+    {
+        // FIXME: Narrowing from using cast because types mismatched with expiry height as uint32, height as int
```
Adding TX_EXPIRY_HEIGHT_THRESHOLD

bitcartel commented on src/gtest/test_mempool.cpp:
```diff
@@ -136,3 +137,80 @@ TEST(Mempool, TxInputLimit) {
+    // CMutableTransaction throws the exception which should only be thrown
+    // during serialization and deserialization, but UpdateHash gets invoked
+    // which performs serialization.
+    // CTransaction tx1(mtx);
```
Ok, keeping the exception throwing.   I already had a unit test which caught this.

More info on "allowed exception"?

bitcartel commented on src/main.cpp:
```diff
@@ -862,16 +870,40 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state,
-    if (tx.nVersion < MIN_TX_VERSION) {
+    // Check min and max tx version of pre-Overwinter and Overwintered transactions.
+    if (tx.fOverwintered) {
+        if (tx.nVersion > CTransaction::OVERWINTER_MAX_CURRENT_VERSION || tx.nVersion < CTransaction::OVERWINTER_MIN_CURRENT_VERSION) {
```
I'm going to leave in CTransaction for now, where MIN/MAX_CURRENT_VERSION are.  At a later date we can move the whole lot over into consensus.h.

bitcartel commented:
Forced push update to address comments.

Issue to resolve:

CheckBlock - how to best determine the block height, or pass in the consensus branch id... per @str4d's comment: https://github.com/zcash/zcash/pull/2925#discussion_r166126880

str4d commented:
Looks like you'll need to rebase on master to remove the upgrade mechanism commits from this PR. GitHub oddity *shrug*

str4d commented on src/main.cpp:
```diff
@@ -862,16 +870,45 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state,
+    }
+
+    // Check Overwintered tx version group id matches what Zcash has assigned to v3 tx format version.
+    if (tx.fOverwintered && tx.nVersion == 3 && tx.nVersionGroupId != V3_VERSION_GROUP_ID) {
```
Check this before the `nExpiryHeight` check, because if `nVersionGroupId` is incorrect then we can't make any assumptions about the contents of `nExpiryHeight`.

str4d commented on src/main.cpp:
```diff
@@ -862,16 +870,45 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state,
+    // with version format >= 3 must have the fOverwintered flag set.
+    // With Sprout consensus rules, where the branch id is 0, we do not need to perform this check as the
+    // Sprout epoch allows for transactions with version number >= 3 (without setting fOverwintered flag).
+    if (NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId == consensusBranchId) {
```
I was going to comment that this might be better as a contextual rule, since we want it to apply for all future upgrades. However, once Overwinter is active, if we have never seen a transaction pre-Overwinter that violates this rule, we could remove this gate and just enforce it everywhere, in which case it requires no context and makes sense here. So we can leave it as-is.

bitcartel replied:
Yep, I spent a while fiddling with this last night, and I think this works for now.

str4d commented on src/gtest/test_mempool.cpp:
```diff
@@ -136,3 +139,35 @@ TEST(Mempool, TxInputLimit) {
 }
 
+// Valid overwinter v3 format tx gets rejected because overwinter hasn't activated yet.
+TEST(Mempool, overwinter_not_active_yet) {
```
Use `CamelCase` (which is what we are supposed to use with Google Test, as underscores could [collide with internal representations of test names](https://github.com/google/googletest/blob/master/googletest/docs/FAQ.md#why-should-not-test-case-names-and-test-names-contain-underscore), so I'm slowly ensuring we convert).

bitcartel replied:
It was camel case but it was hard to read so I changed it.  Will change back.

str4d commented on src/gtest/test_mempool.cpp:
```diff
@@ -136,3 +139,35 @@ TEST(Mempool, TxInputLimit) {
+}
+
+// Sprout tx version 3 gets past overwinter format checks present in CheckTransactionWithoutProofVerification
+TEST(Mempool, not_overwinter_v3) {
```
`CamelCase`

arielgabizon commented:
This should be rebased on https://github.com/zcash/zcash/pull/2898 to include only the last commit

str4d commented on src/gtest/test_mempool.cpp:
```diff
@@ -136,3 +139,35 @@ TEST(Mempool, TxInputLimit) {
 
+// Valid overwinter v3 format tx gets rejected because overwinter hasn't activated yet.
+TEST(Mempool, overwinter_not_active_yet) {
+    SelectParams(CBaseChainParams::TESTNET);
```
Non-blocking: This test will fail if Overwinter is ever set to `ALWAYS_ACTIVE` on testnet. I doubt we will do so (only case I can imagine is some custom private testnet), but downstream code forks might want to. Not too concerned however because doing this would also require changes to the genesis block.

bitcartel replied:
Added comment about this to test.

str4d commented on src/gtest/test_mempool.cpp:
```diff
@@ -92,6 +93,8 @@ TEST(Mempool, PriorityStatsDoNotCrash) {
 }
 
 TEST(Mempool, TxInputLimit) {
+    SelectParams(CBaseChainParams::TESTNET);
```
Why is this necessary?

bitcartel replied:
The tx fails CheckValidation when calling AcceptToMemoryPool, which is the intended behaviour. However, because we now have to pass in a branch id to CheckValidation, Params().GetConsensus() is invoked *before* calling CheckValidation, and Params() hasn't been set up yet, so then we get an error.

str4d commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -391,3 +392,155 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+    CDataStream ss(SER_DISK, PROTOCOL_VERSION);
+    ss << mtx;
+    CTransaction tx;
+    ss >> tx;
```
This checks serialization of `CMutableTransaction` and deserialization of `CTransaction`. We also need to check the reverse. Maybe the other two combinations as well (`mtx -> mtx` and `tx -> tx`) for completeness?

arielgabizon commented on src/main.cpp:
```diff
@@ -646,11 +646,19 @@ unsigned int LimitOrphanTxSize(unsigned int nMaxOrphans) EXCLUSIVE_LOCKS_REQUIRE
-        reason = "version";
-        return false;
+    if (consensusBranchId == NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId ) {
+        if (tx.nVersion > CTransaction::OVERWINTER_MAX_CURRENT_VERSION || tx.nVersion < CTransaction::OVERWINTER_MIN_CURRENT_VERSION) {
```
The name of these constants seems a bit ad hoc. Wondering if it's better they be fields in the NetworkUpgrade object.

bitcartel replied:
Keeping them in CTransaction for now like existing CTransaction::MIN_VERSION.  Easy for us to move/rename later.

str4d commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -391,3 +392,155 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+    mtx.nExpiryHeight = 0;
+    CTransaction tx(mtx);
+    MockCValidationState state;
+    EXPECT_TRUE(CheckTransactionWithoutProofVerification(tx, state, NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId));
```
Wrap long line.

str4d commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -391,3 +392,155 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+    {
+        CTransaction tx(mtx);
+        MockCValidationState state;
+        EXPECT_TRUE(CheckTransactionWithoutProofVerification(tx, state, NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId));
```
Wrap long line.

str4d commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -391,3 +392,155 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+        mtx.nExpiryHeight = TX_EXPIRY_HEIGHT_THRESHOLD;
+        CTransaction tx(mtx);
+        MockCValidationState state;
+        EXPECT_TRUE(CheckTransactionWithoutProofVerification(tx, state, NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId));
```
Wrap long line.

str4d commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -391,3 +392,155 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+        CTransaction tx(mtx);
+        MockCValidationState state;
+        EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "overwinter-expiry-height-too-high", false)).Times(1);
+        CheckTransactionWithoutProofVerification(tx, state, NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId);
```
Wrap long lines.

str4d commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -391,3 +392,155 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+        CTransaction tx(mtx);
+        MockCValidationState state;
+        EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "overwinter-expiry-height-too-high", false)).Times(1);
+        CheckTransactionWithoutProofVerification(tx, state, NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId);
```
Wrap long lines.

arielgabizon commented on src/main.cpp:
```diff
@@ -862,16 +870,45 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state,
+    }
+
+    // Check Overwintered tx version group id matches what Zcash has assigned to v3 tx format version.
+    if (tx.fOverwintered && tx.nVersion == 3 && tx.nVersionGroupId != V3_VERSION_GROUP_ID) {
```
`nVersionGroupId` not discussed in ZIP

str4d replied:
For context in the meantime, it replaces the format branch ID.

bitcartel replied:
@arielgabizon will update zip soon

arielgabizon commented on src/primitives/transaction.h:
```diff
@@ -351,11 +366,39 @@ class CTransaction
+            uint32_t header = this->nVersion;
+            // When serializing Overwintered tx, the 4 byte header is the combination of fOverwintered and nVersion
+            if (fOverwintered) {
+                header |= 1 << 31;
```
Does this ever change `header`? Seems you're setting the 31st bit only when it was set already.

str4d replied:
`this->nVersion` (the internal representation of the version number) will now always positive, and thus will never have the 31st bit set.

arielgabizon replied:
OK. What I meant was that for overwinter the value of header in 372 is like the final value,
but I missed that setting it this way gives the correct value also for pre-overwinter

bitcartel commented on src/main.cpp:
```diff
@@ -646,11 +646,19 @@ unsigned int LimitOrphanTxSize(unsigned int nMaxOrphans) EXCLUSIVE_LOCKS_REQUIRE
 
 
-bool IsStandardTx(const CTransaction& tx, string& reason)
+bool IsStandardTx(const CTransaction& tx, string& reason, uint32_t consensusBranchId)
```
I think this can be refactor later on as currently quite a few places pass in the reason parameter.  If we changed it now, a whole bunch of code elsewhere would change and it's probably better to keep the Overwinter changes as small as possible for now

bitcartel commented on src/main.cpp:
```diff
@@ -1074,18 +1111,51 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
+
+    // Get consensus branch id for the next block, which is what mempool txs want to be mined into.
+    uint32_t consensusBranchId = CurrentEpochBranchId(currentHeight + 1, params); 
+;
```
I changed to the 'Hack source code' font last night... I blame that!  https://sourcefoundry.org/hack/

bitcartel commented on src/main.cpp:
```diff
@@ -1074,18 +1111,51 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
+
+    // Contextual check: If Overwinter is active, reject transactions which are not Overwintered
+    if (isOverwinterActive && !tx.fOverwintered) {
+        return state.DoS(100, error("AcceptToMemoryPool: pre-overwinter tx not allowed"),
```
I agree, I will lower to 10.  Runing `git grep state.DoS` across the repo is interesting, a lot of `100`.

bitcartel commented on src/main.cpp:
```diff
@@ -1074,18 +1111,51 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
+            return state.DoS(100, error("AcceptToMemoryPool: tx expired at block height: %u", tx.nExpiryHeight),
+                             REJECT_INVALID, "tx-expired");
+        }
+    }
```
Removed

bitcartel commented on src/primitives/transaction.h:
```diff
@@ -304,6 +304,16 @@ class CTxOut
+// to a tx format version or group of soft-forking tx format versions.
+// This helps nodes disambiguate between branches using the same version number
+// but where the transaction format itself is different.
+// Forks and clones of Zcash should change this value.
```
Will move to ZIP. I was thinking of a clone at the activation height who wanted their own v3 format, rather than after the activation height. Although they could start a brand new chain using the code, in which case they should change it even if the same v3 format to help disambiguate.

bitcartel commented:
Forced pushed to address review comments.
Added FIXME to ProcessNewBlock to check the consensus branch id is being generated correctly.
Also updated CheckBlock test.

str4d commented on src/main.cpp:
```diff
@@ -862,16 +870,50 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state,
 
-    // Check transaction version
-    if (tx.nVersion < MIN_TX_VERSION) {
+    // Check min and max tx version of pre-Overwinter and Overwinter transactions.  Range allows for soft fork.
```
Defining a range allows for soft-forked transaction versions, but the current values of the Overwinter range endpoints do not. Is there a clearer way to phrase this?

str4d commented on src/main.cpp:
```diff
@@ -1074,8 +1116,32 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
+
+    // Has Overwinter upgrade activated?  Gated to avoid assertion in NetworkUpgradeActive when no active chain e.g. testing.
+    bool isOverwinterActive = false;
+    if (currentHeight > 0) {
```
`if (currentHeight >= 0) {`

str4d commented on src/main.cpp:
```diff
@@ -3131,6 +3201,19 @@ bool ContextualCheckBlock(const CBlock& block, CValidationState& state, CBlockIn
+            return state.DoS(100, error("%s: contains pre-overwinter tx", __func__),
+                            REJECT_INVALID, "cb-no-pre-overwinter-tx-when-active");
+        }
+        if (!isOverwinterActive && tx.fOverwintered) {
```
Sanity-check: this is a new rejection rule that will be enabled on current mainnet (before Overwinter activation) when 1.1.0 is released, but it changes nothing because it is semantically equivalent to the `tx.nVersion < MIN_TX_VERSION` rejection rule.

bitcartel replied:
Right.. so I will remove this as it is redundant... and then gate the entire foreach loop so we are not introducing an Overwinter check (rejection rule) before Overwinter activates.  I'll also remove the test for `cb-no-pre-overwinter-tx-when-active`.

str4d replied:
Hmm, it occurred to me this morning that technically this isn't entirely redundant, because we've changed how we parse transactions, storing the sign bit in `tx.fOverwintered`. So the previous `tx.nVersion < MIN_TX_VERSION` rule is now semantically equivalent to `tx.fOverwintered || tx.nVersion < MIN_TX_VERSION`. Maybe we should change that rule instead?

str4d replied:
i.e. what we actually do is move the above rule over next to the existing rule and combine them.

str4d replied:
Or, take the `GetHeader()` function that I extracted in #2903 and have it return an `int32_t`, and then just use `tx.GetHeader() < MIN_TX_VERSION` here.

str4d replied:
Per my comment in https://github.com/zcash/zcash/pull/2925/files#r167231233 we cannot remove this rule, as it is not in fact redundant. We also need to re-check all consensus rules to ensure that the semantic change of `tx.nVersion` (to never have its sign bit set) doesn't alter any other consensus rules (I'm reasonably confident that `tx.nVersion < MIN_TX_VERSION` is the only one affected by this, but I want to be certain).

bitcartel replied:
@daira @str4d This is a good argument for having a control byte (with the msb set) so that the existing `nVersion` field does not change semantics.  Per comment here: https://github.com/zcash/zips/pull/133#issuecomment-361432773
We could still do this...

str4d commented on src/main.cpp:
```diff
@@ -3305,9 +3391,17 @@ static bool IsSuperMajority(int minVersion, const CBlockIndex* pstart, unsigned
+    {
+        LOCK(cs_main);
+        // FIXME: IS THIS THE RIGHT HEIGHT TO USE?
+        height = chainActive.Height() + 1;
```
Ooh, this is a problem. I don't believe we can assume that the block being processed would end up at the chain tip; we don't know its height until after `AcceptBlock` is called below. This means that we cannot check any consensus rules in `CheckBlock` that require knowledge of the block's height, and thus we can never pass `consensusBranchId` in. We should therefore remove `consensusBranchId` as an argument to `CheckBlock`, `CheckTransaction` and `CheckTransactionWithoutProofVerification`.

This affects two rules:

- https://github.com/zcash/zcash/pull/2925/files#r166610870 which I was already on the fence about, so that means we just need to move that rule out of `CheckTransaction` for now, and can move it back after Overwinter.
- The `joinSplitSig` check in `CheckTransaction`, which in #2903 needs to know `consensusBranchId` from Overwinter.

The one annoyance is that there is currently no common place where general contextual checks are made on transactions, so we will need to move them into both `AcceptToMemoryPool` and `ContextualCheckBlock`. It might make sense to add a new function to contain them, assuming that we want the same DoS level for both mempool and blockchain. Upstream AFAIR hasn't refactored these kinds of rules yet, probably for this reason.

str4d replied:
I've just grepped around, and `ContextualCheckBlock` is called from `AcceptBlock` (i.e. `ProcessNewBlock`) and `TestBlockValidity`, so that means we should still always validate the above rules when a new block is received. It *isn't* called from `VerifyDB` (only `CheckBlock` and `ContextualCheckInputs` are triggered from there), but that should be fine, as it's only a consistency check.

str4d commented on src/main.cpp:
```diff
@@ -3345,11 +3439,13 @@ bool TestBlockValidity(CValidationState &state, const CBlock& block, CBlockIndex
     // JoinSplit proofs are verified in ConnectBlock
     auto verifier = libzcash::ProofVerifier::Disabled();
+    // Get branch id for the consensus rules a new block at current tip +1 should be checked against.
+    uint32_t consensusBranchId = CurrentEpochBranchId(indexDummy.nHeight, Params().GetConsensus());
```
No longer necessary, per above.

bitcartel replied:
Removed

str4d commented on src/primitives/transaction.h:
```diff
@@ -351,11 +362,39 @@ class CTransaction
+
+        bool isOverwinterV3 = fOverwintered &&
+                              nVersion == 3 &&
+                              nVersionGroupId == OVERWINTER_VERSION_GROUP_ID;
```
Swap `nVersionGroupId` and `nVersion` checks around to match precedence.

str4d commented on src/primitives/transaction.h:
```diff
@@ -425,11 +467,39 @@ struct CMutableTransaction
+
+        bool isOverwinterV3 = fOverwintered &&
+                              nVersion == 3 &&
+                              nVersionGroupId == OVERWINTER_VERSION_GROUP_ID;
```
Ditto.

str4d commented on src/main.cpp:
```diff
@@ -862,16 +870,50 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state,
+                REJECT_INVALID, "bad-tx-overwinter-version-too-high");
+        }
+    }
+    else if (tx.nVersion < MIN_TX_VERSION) {
```
Hmm, it occurs to me that this change here is technically a change to the current consensus rules, because if this function is used anywhere on its own, it changes the allowed version range from `tx.header >= MIN_TX_VERSION` to `tx.header >= MIN_TX_VERSION || tx.header == -3` (using `tx.header` as a stand-in for what used to be internally named `tx.nVersion`. This means we need to be very careful to ensure we have somewhere a restriction `!fOverwinterActive && tx.header < MIN_TX_VERSION`.

bitcartel replied:
In `CheckTransactionWithoutProofVerification`

bitcartel commented:
Force pushed to address latest round of review comments.  Still to do:

- [ ] Create v3 (transparent and shielded) transactions, when Overwinter is active
- [ ] RPC commands return v3 transaction info
- [ ] QA tests
- [ ] More unit tests where applicable

str4d commented on src/main.h:
```diff
@@ -332,6 +333,10 @@ bool ContextualCheckInputs(const CTransaction& tx, CValidationState &state, cons
 
+/** Check a transaction contextually against a set of consensus rules */
+bool ContextualCheckTransaction(const CTransaction& tx, CValidationState &state,
+                                int nHeight = 0, int dosLevel = 0);
```
Is there anywhere we don't want to pass in a height? DoS level I can understand not passing in (in tests).

bitcartel replied:
I'm going to remove both defaults.  The caller must provide the context which also includes dos level.

str4d commented on src/main.cpp:
```diff
@@ -5735,3 +5825,19 @@ class CMainCleanup
+    CMutableTransaction mtx;
+
+    bool isOverwinter = NetworkUpgradeActive(nHeight, consensusParams, Consensus::UPGRADE_OVERWINTER);
+    if (isOverwinter) {
```
Nit: `isOverwintered`.

str4d commented on src/main.cpp:
```diff
@@ -3345,7 +3436,6 @@ bool TestBlockValidity(CValidationState &state, const CBlock& block, CBlockIndex
     indexDummy.nHeight = pindexPrev->nHeight + 1;
     // JoinSplit proofs are verified in ConnectBlock
     auto verifier = libzcash::ProofVerifier::Disabled();
-
```
Leave this line in.

str4d replied:
I'm still here!

str4d commented on src/main.cpp:
```diff
@@ -1074,6 +1151,14 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
+
+    // FIXME: DoS level set to 10 to be more forgiving... or use 100?
+    // Check transaction contextually against the set of consensus rules which apply in the next block to be mined.
+    if (!ContextualCheckTransaction(tx, state, nextBlockHeight, 10)) {
```
Do the contextual checks after the non-contextual ones (i.e. put this in between `CheckTransaction()` and `IsCoinBase()`).

bitcartel replied:
Updated mempool test for this and added new unit tests.

str4d commented on src/main.cpp:
```diff
@@ -840,6 +846,53 @@ unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& in
+ * Notes:
+ * 1. AcceptToMemoryPool calls this function and CheckTransaction.
+ * 2. ProcessNewBlock calls AcceptBlock which calls CheckBlock (which calls CheckTransaction)
+ *    and ContextualCheckBlock which calls this function.
```
For symmetry: `and ContextualCheckBlock (which calls this function).`

str4d commented on src/main.cpp:
```diff
@@ -840,6 +846,53 @@ unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& in
+                         REJECT_INVALID, "tx-overwinter-version-too-high");
+    }
+
+    // If Overwinter rules apply, reject transactions intended for Sprout
```
May as well move this up to below the equivalent Sprout rule.

bitcartel commented:
Force pushed for latest review comments.

str4d commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -391,3 +391,346 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+        FAIL() << "Expected std::ios_base::failure";
+    }
+    catch(std::ios_base::failure & err) {
+        EXPECT_EQ(err.what(), std::string("Unknown transaction format: iostream error"));
```
`EXPECT_THAT(err.what(), HasSubstr(std::string("Unknown transaction format")))` (see the [CheatSheet](https://github.com/google/googletest/blob/master/googlemock/docs/CheatSheet.md) for details).

bitcartel replied:
Bookmarked!

str4d commented on src/gtest/test_mempool.cpp:
```diff
@@ -9,6 +10,9 @@
 #include "policy/fees.h"
 #include "util.h"
 
+// Implementaion is in test_checktransaction.cpp
```
`Implementation`

str4d commented on src/main.cpp:
```diff
@@ -641,16 +641,22 @@ unsigned int LimitOrphanTxSize(unsigned int nMaxOrphans) EXCLUSIVE_LOCKS_REQUIRE
+    if (isOverwinter) {
+        // Overwinter consensus rules apply
+        if (tx.nVersion > CTransaction::OVERWINTER_MAX_CURRENT_VERSION || tx.nVersion < CTransaction::OVERWINTER_MIN_CURRENT_VERSION) {
+            reason = "invalid-overwinter-version";
```
`s/invalid-//` (in the context of this function it's not technically invalid, just the reason for it not being standard.)

str4d commented on src/main.cpp:
```diff
@@ -840,6 +846,53 @@ unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& in
+ * 
+ * Notes:
+ * 1. AcceptToMemoryPool calls this function and CheckTransaction.
+ * 2. ProcessNewBlock calls AcceptBlock which calls CheckBlock (which calls CheckTransaction)
```
`ProcessNewBlock calls AcceptBlock, which calls`

str4d commented on src/main.cpp:
```diff
@@ -840,6 +846,53 @@ unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& in
+ * Check a transaction contextually against a set of consensus rules valid at a given block height.
+ * 
+ * Notes:
+ * 1. AcceptToMemoryPool calls this function and CheckTransaction.
```
`calls CheckTransaction and this function.`

str4d commented on src/main.cpp:
```diff
@@ -866,12 +919,36 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+        }
+    }
+
+    if (!tx.fOverwintered && tx.nVersion < MIN_TX_VERSION) {
```
Add comment about how the previous consensus rule `tx.nVersion < MIN_TX_VERSION` is now split between here and `ContextualCheckTransaction`.

bitcartel replied:
Is it? I was writing a comment but then wasn't sure what you meant.  We'll have to add later after discussion.

str4d replied:
This ties into the other comments about the fact that the parser now ensures that `tx.nVersion` will never be negative, so this check which previously excluded negative values now is split between here (which excludes zero) and the `fOverwintered` check in `ContextualCheckTransaction` (which excludes what Sprout considers negative).

str4d commented on src/main.cpp:
```diff
@@ -1078,14 +1155,21 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
     if (!CheckTransaction(tx, state, verifier))
         return error("AcceptToMemoryPool: CheckTransaction failed");
 
+    // FIXME: DoS level set to 10 to be more forgiving... or use 100?
```
I think we want to be forgiving here (as mempool transactions could race with the activation block), but not *too* forgiving (for DoS prevention).

daira replied:
Delete "FIXME:" and "... or use 100?".

str4d commented on src/main.cpp:
```diff
@@ -3257,6 +3347,7 @@ bool AcceptBlock(CBlock& block, CValidationState& state, CBlockIndex** ppindex,
 
     // See method docstring for why this is always disabled
     auto verifier = libzcash::ProofVerifier::Disabled();
+
```
Remove this line (it's not the one that was removed below 😜).

str4d commented:
@zkbot try

ebfull commented on src/main.cpp:
```diff
@@ -641,16 +641,22 @@ unsigned int LimitOrphanTxSize(unsigned int nMaxOrphans) EXCLUSIVE_LOCKS_REQUIRE
+            return false;
+        }
+    } else {
+        // Sprout consensus rules apply
```
`IsStandard` does not currently check if the transaction is `overwintered`, and so it would return `true` for an `overwintered` transaction with version number 2 before Overwinter is activated -- which is invalid as per `ContextualCheckTransaction`.

Is this okay? It doesn't sound like it, unless every place where `IsStandard` applies, `ContextualCheckTransaction` applies also. Example: if this weren't the case in `AcceptToMemoryPool`, we would accept a transaction into the mempool which violates `ContextualCheckTransaction`.  (`ContextualCheckTransaction` does in fact get called by `AcceptToMemoryPool`, but what other places do we call `IsStandard`?)

str4d replied:
`IsStandard` is a local restriction that is strictly a tightening of the consensus rules (so yes, `ContextualCheckTransaction` should always be called before `IsStandard`), and is only enforced in `AcceptToMemoryPool` (otherwise its rules would be checked on actual blocks, which would make them consensus rules instead of standard rules).

str4d commented:
- The btest failure should be fixed by https://github.com/zcash/zcash/pull/2925#discussion_r167580249.
- The gtest failure can be fixed by cherry-picking https://github.com/zcash/zcash/pull/2903/commits/7abb30ac6980115b12543011e7c9647ad12856ab from #2903.

daira commented on src/main.cpp:
```diff
@@ -866,12 +919,36 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
     }
 
+    // Check Overwintered tx version group id matches what Zcash has assigned to v3 tx format version.
+    if (tx.fOverwintered && tx.nVersion == 3 && tx.nVersionGroupId != OVERWINTER_VERSION_GROUP_ID) {
```
I know ``OVERWINTER_MAX_CURRENT_VERSION`` is set to 3 so it doesn't matter for this upgrade, but it still seems semantically wrong for this to be conditional on ``tx.nVersion == 3``. The version group ID should be a recognized value, for any version in the range ``${UPGRADE}_MIN_CURRENT_VERSION`` to ``${UPGRADE}_MAX_CURRENT_VERSION``. Since this code is likely to be copied for future upgrades, I'd prefer the more restrictive condition (i.e. just ``(tx.fOverwintered && tx.nVersionGroupId != OVERWINTER_VERSION_GROUP_ID)``).

daira commented on src/main.h:
```diff
@@ -342,7 +346,7 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
  * @return True if all outputs (scriptPubKeys) use only standard transaction forms
  */
-bool IsStandardTx(const CTransaction& tx, std::string& reason);
+bool IsStandardTx(const CTransaction& tx, std::string& reason, int nHeight = 0);
```
I'm wary of this default. Shouldn't all uses now pass the height? (I know this involves updating a shedload of tests :-/)

bitcartel replied:
Yes, that's the main reason for this.  Can remove default and update tests later.

daira replied:
Please file a ticket.

daira commented on src/primitives/transaction.cpp:
```diff
@@ -163,19 +164,23 @@ void CTransaction::UpdateHash() const
+    *const_cast<uint32_t*>(&nExpiryHeight) = tx.nExpiryHeight;
     *const_cast<std::vector<JSDescription>*>(&vjoinsplit) = tx.vjoinsplit;
     *const_cast<uint256*>(&joinSplitPubKey) = tx.joinSplitPubKey;
     *const_cast<joinsplit_sig_t*>(&joinSplitSig) = tx.joinSplitSig;
```
Please let's fix the ``const_cast`` thing before 1.1.0.

daira commented on src/primitives/transaction.h:
```diff
@@ -319,9 +321,10 @@ class CTransaction
 
-    // Transactions that include a list of JoinSplits are version 2.
+    // Transactions that include a list of JoinSplits are >= version 2.
     static const int32_t MIN_CURRENT_VERSION = 1;
```
Yes, I already suggested changing these to ``SPROUT_{MIN,MAX}_CURRENT_VERSION`` above.

Also, is the ``CURRENT_`` applicable any more?

daira commented on src/primitives/transaction.h:
```diff
@@ -319,9 +326,11 @@ class CTransaction
+    static const int32_t OVERWINTER_MAX_CURRENT_VERSION = 3;
 
     static_assert(MIN_CURRENT_VERSION >= MIN_TX_VERSION,
                   "standard rule for tx version should be consistent with network rule");
```
Add ``static_assert(OVERWINTER_MIN_CURRENT_VERSION >= MIN_TX_VERSION, ...);``.

daira commented on src/primitives/transaction.h:
```diff
@@ -425,11 +468,39 @@ struct CMutableTransaction
         READWRITE(nLockTime);
+        if (isOverwinterV3) {
+            READWRITE(nExpiryHeight);
+        }
```
All the near-duplicate code here bothers me.

str4d replied:
It will be fixed at the same time the `const_cast` stuff is.

bitcartel replied:
@daira I can't add a comment to your comment above.  No, exceptions during deserialization are rare but there are classes that do this.  I found between 5-10 across the code base, example:
```
template <typename Stream, typename Operation>
    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
        unsigned char leadingByte = 0x00;
        READWRITE(leadingByte);

        if (leadingByte != 0x00) {
            throw std::ios_base::failure("lead byte of NotePlaintext is not recognized");
        }

        READWRITE(value);
        READWRITE(rho);
        READWRITE(r);
        READWRITE(memo);
    }
```

str4d replied:
I keep writing comments about this and then being unable to find them...

Exceptions are not rare in terms of parser coverage, because `ReadCompactSize` throws on non-canonical encodings, so every single serialization type that uses a compact size (string, vector, map, set, ...) has the ability to throw.

daira replied:
In that case it's fine to throw here.

str4d commented on src/gtest/test_checkblock.cpp:
```diff
@@ -75,3 +107,91 @@ TEST(ContextualCheckBlock, BadCoinbaseHeight) {
+    EXPECT_TRUE(ContextualCheckBlock(block, state, &indexPrev));
+
+    // Revert to default
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_OVERWINTER, Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT);
```
This isn't a Boost test, so you'd need to do this the same way I did for the upgrade tests (with a testing subclass).

daira replied:
``BOOST_SCOPE_EXIT`` is completely general and doesn't depend on any test framework. It's just a way of simulating try..finally.

daira replied:
Filed #2953.

str4d commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -391,3 +391,346 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+        CTransaction tx(mtx);
+
+        // Evil
+        *const_cast<int32_t*>(&tx.nVersion) = CTransaction::OVERWINTER_MAX_CURRENT_VERSION + 1;
```
> No it's not! It makes the test have undefined behaviour which means, for example, that it could spuriously pass.

Okay, fair enough: for this specific class of evil, it's possible for the test to spuriously pass. But then, is that worse than not having a test at all (which is functionally equivalent to a test that always passes), if the evil is sufficiently documented?

> Also, I disagree in principle that it's necessary to test object states (for constant objects) that the constructor rejects.

The only reason the constructor rejects it is because the constructor calls `UpdateHash()` which performs a serialization. So it's not quite as direct as that, but I can see the reasoning.

daira replied:
Yes, it is worse. I don't want to have to think about undefined behaviour in the test suite. It complicates reasoning about what test results tell you.

daira replied:
In general we should have zero tolerance for undefined behaviour. We've been too lax about this :-/

bitcartel commented:
- Pushed a new commit for reviewers @ebfull @arielgabizon to focus on creation of v3 transactions from call sites where v1 and v2 tx are currently created (need a qa test as well, but no need to block on it, and there are some other tests related to this)
- @daira I read all your comments and will address in a separate commit as I didn't want to block other reviewers.
- @arcalinea Please rebase your tx expiry, thx.
- @str4d We can squash commits later before we merge, would have preferred to force push, but the size of the PR is getting bigger and harder for reviewers to keep track of new changes

str4d commented on src/test/rpc_wallet_tests.cpp:
```diff
@@ -1320,8 +1344,14 @@ BOOST_AUTO_TEST_CASE(rpc_z_shieldcoinbase_internals)
+    // Mutable tx containing contextual information we need to build tx
+    UniValue retValue = CallRPC("getblockcount");
+    int nHeight = retValue.get_int();
+    CMutableTransaction mtx = CreateNewContextualCMutableTransaction(Params().GetConsensus(), 1 + nHeight);
```
`nHeight + 1`

str4d commented on src/wallet/asyncrpcoperation_sendmany.cpp:
```diff
@@ -50,14 +50,16 @@ int find_output(UniValue obj, int n) {
-        fromaddress_(fromAddress), t_outputs_(tOutputs), z_outputs_(zOutputs), mindepth_(minDepth), fee_(fee), contextinfo_(contextInfo)
+        tx_(contextualTx), fromaddress_(fromAddress), t_outputs_(tOutputs), z_outputs_(zOutputs), mindepth_(minDepth), fee_(fee), contextinfo_(contextInfo)
 {
+    assert(contextualTx.nVersion >= 2);  // transaction format version must support vjoinsplit
```
A side-effect of this is that `z_sendmany` no longer creates v1 transactions for `t->t`. This should be fine though, as anyone who has made the effort to use `z_sendmany` should already support parsing v2 transactions.

arielgabizon replied:
Is this addition related to NU0?

daira replied:
This is technically violating a SHOULD in the spec: "Software that creates transactions SHOULD use version 1 for transactions with no JoinSplit descriptions."

Shall I remove this from the spec?

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3525,9 +3525,17 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
+    // Contextual transaction we will build on
+    CMutableTransaction contextualTx = CreateNewContextualCMutableTransaction(
+        Params().GetConsensus(),
+        1 + chainActive.Height());
```
`chainActive.Height() + 1`

str4d replied:
Looks like this is still unchanged in this particular spot.

bitcartel replied:
Specsavers for me.

str4d commented on src/main.cpp:
```diff
@@ -5838,6 +5838,9 @@ CMutableTransaction CreateNewContextualCMutableTransaction(const Consensus::Para
         // Expiry height is not set. Only fields required for a parser to treat as a valid Overwinter V3 tx.
+
+        // TODO: In future, when moving from Overwinter to Sapling, it will be useful
+        // for the expiry height to be set to min(blocks_until_upgrade, default_expiry_height)
```
`min(activation_height - 1, default_expiry_height)` (since `nExpiryHeight` is the last mineable height).

bitcartel commented on src/gtest/test_checkblock.cpp:
```diff
@@ -75,3 +107,91 @@ TEST(ContextualCheckBlock, BadCoinbaseHeight) {
+
+
+
+// Test block evaluated under Overwinter rules will reject Sprout transactions
```
Good idea.  Added a test to be explicit.

bitcartel commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -391,3 +391,346 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+        CTransaction tx(mtx);
+
+        // Evil
+        *const_cast<int32_t*>(&tx.nVersion) = CTransaction::OVERWINTER_MIN_CURRENT_VERSION - 1;
```
Sorry about the evil but exception gets thrown in serialization code.

bitcartel commented on src/gtest/test_mempool.cpp:
```diff
@@ -136,3 +142,93 @@ TEST(Mempool, TxInputLimit) {
+    CMutableTransaction mtx = GetValidTransaction();
+    mtx.vjoinsplit.resize(0); // no joinsplits
+    mtx.fOverwintered = false;
+    mtx.nVersion = -3;
```
It maintains existing v1 and v2 behaviour where you can set nVersion to a negative value.
```
            // When serializing v1 and v2, the 4 byte header is nVersion
            uint32_t header = this->nVersion;
            // When serializing Overwintered tx, the 4 byte header is the combination of fOverwintered and nVersion
            if (fOverwintered) {
                header |= 1 << 31;
            }
            READWRITE(header);
```

str4d replied:
Hmm, @daira is right here. The old `nVersion` can be negative, but we now internally have `nVersion` always be positive; a negative internal `nVersion` should be considered an error state. IF we want to ensure that negative transactions are rejected under Sprout, we need to be checking `!isOverwinterActive && fOverwintered == true`.

This is the kind of thing that makes me wonder if we should just make `nVersion` a `uint32_t` now, to be explicit about the fact that the negative bit does not exist internally any more, but is explicitly stored in `fOverwintered`.

arielgabizon replied:
This has to do with the fact that when we serialize and `fOverWintered==true` we only take 31 bits from `nVersion`, iiuc, which is a bit tricky, cause serialization is just supposed to serialize. I wonder what's the potential of that leading to problems at some point

str4d replied:
Also, if the intent of this test is to emulate a parser failure to mask the sign off of a Sprout transaction with a version that collides with Overwinter, then it should use the int-coerced negative value that Overwinter is "using", i.e. `static_cast<int>((1 << 31) | 3)`.

bitcartel commented on src/main.cpp:
```diff
@@ -641,16 +641,22 @@ unsigned int LimitOrphanTxSize(unsigned int nMaxOrphans) EXCLUSIVE_LOCKS_REQUIRE
+        }
+    } else {
+        // Sprout consensus rules apply
+        if (tx.nVersion > CTransaction::MAX_CURRENT_VERSION || tx.nVersion < CTransaction::MIN_CURRENT_VERSION) {
```
Good idea. I've renamed 'SPROUT_MIN_CURRENT_VERSION' and 'SPROUT_MAX_CURRENT_VERSION' to distinguish with Overwinter. I didn't use`MIN_UPGRADE` as I found that more confusing.

bitcartel commented on src/main.cpp:
```diff
@@ -840,6 +846,53 @@ unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& in
+    // If Overwinter rules apply, reject transactions with valid version but missing overwinter flag
+    if (isOverwinter && tx.nVersion >= CTransaction::OVERWINTER_MIN_CURRENT_VERSION && !tx.fOverwintered) {
+        return state.DoS(dosLevel, error("ContextualCheckTransaction(): overwinter flag must be set"),
+                         REJECT_INVALID, "tx-overwinter-flag-not-set");
```
I don't know and will need to research, maybe @str4d or @ebfull knows off the top of their head?

str4d replied:
Default ban time is 24 hours, and the misbehaving node handler uses the default.

arielgabizon commented on src/primitives/transaction.h:
```diff
@@ -319,22 +326,30 @@ class CTransaction
-    static const int32_t MIN_CURRENT_VERSION = 1;
-    static const int32_t MAX_CURRENT_VERSION = 2;
+    // Transactions that include a list of JoinSplits are >= version 2.
+    static const int32_t SPROUT_MIN_CURRENT_VERSION = 1;
```
`SPROUT_CURRENT_MIN_VERSION`  or `SPROUT_MIN_VERSION` sound clearer to me than `SPROUT_MIN_CURRENT_VERSION`

bitcartel commented:
Updated for lastest review comments. Added comment in main.cpp about `tx.nVersion < MIN_TX_VERSION`.

I've kept the 'evil' tests in for two reasons:

1. Review commens is _"The reason is that the parser rules are consensus rules themselves"_ however this has not been the case historically since the parser allowed negative values in the version field, which breaks the consensus rule `< MIN_TX`
2. White box testing of `ContextualCheckTransaction` and `CheckTransactionWithoutProofVerification` for bad version numbers gives us higher assurance.  That is, if there is bug in the parser, we have tests to confirm that those transactions won't get accepted into the memory pool.

arcalinea commented:
Build on this branch fails with this error, discovered after I rebased tx expiry on top:
```
gtest/zcash_gtest-test_tautology.o: In function `__static_initialization_and_destruction_0':
/home/jay/Zcash/zcash/src/gtest/test_tautology.cpp:3: undefined reference to `testing::internal::MakeAndRegisterTestInfo(char const*, char const*, char const*, char const*, void const*, void (*)(), void (*)(), testing::internal::TestFactoryBase*)'
/home/jay/Zcash/zcash/src/gtest/test_tautology.cpp:7: undefined reference to `testing::internal::MakeAndRegisterTestInfo(char const*, char const*, char const*, char const*, void const*, void (*)(), void (*)(), testing::internal::TestFactoryBase*)'
gtest/zcash_gtest-test_metrics.o: In function `__static_initialization_and_destruction_0':
/home/jay/Zcash/zcash/src/gtest/test_metrics.cpp:7: undefined reference to `testing::internal::MakeAndRegisterTestInfo(char const*, char const*, char const*, char const*, void const*, void (*)(), void (*)(), testing::internal::TestFactoryBase*)'
/home/jay/Zcash/zcash/src/gtest/test_metrics.cpp:55: undefined reference to `testing::internal::MakeAndRegisterTestInfo(char const*, char const*, char const*, char const*, void const*, void (*)(), void (*)(), testing::internal::TestFactoryBase*)'
/home/jay/Zcash/zcash/src/gtest/test_metrics.cpp:96: undefined reference to `testing::internal::MakeAndRegisterTestInfo(char const*, char const*, char const*, char const*, void const*, void (*)(), void (*)(), testing::internal::TestFactoryBase*)'
gtest/zcash_gtest-test_random.o:/home/jay/Zcash/zcash/src/gtest/test_random.cpp:8: more undefined references to `testing::internal::MakeAndRegisterTestInfo(char const*, char const*, char const*, char const*, void const*, void (*)(), void (*)(), testing::internal::TestFactoryBase*)' follow
```

str4d commented:
> Review comment is *"The reason is that the parser rules are consensus rules themselves"* however this has not been the case historically since the parser allowed negative values in the version field, which breaks the consensus rule `< MIN_TX_VERSION`.

Just to be clear, it has always been the case that the parser rules are consensus rules, and that not all conensus rules were enumerated in `main.cpp` (e.g. prohibiting non-canonical CompactSize fields). I agree it's the case that this particular rule was not previously enforced in the parser, and that as a whole, the enforced consensus rules are a subset of the parser rules.

str4d commented:
> White box testing of `ContextualCheckTransaction` and `CheckTransactionWithoutProofVerification` for bad version numbers gives us higher assurance. That is, if there is bug in the parser, we have tests to confirm that those transactions won't get accepted into the memory pool.

I'm fine with this justification (with a view to remove in future once we are confident in the parser), as long as any undefined behaviour that might occur as a result of the evil is limited to that test. @daira asserts that [undefined behaviour in one test puts into question the results of all of them](https://github.com/zcash/zcash/pull/2925#discussion_r167723073). Is there a way we could sandbox the tests? Maybe run them in a forked process, or tweak CI to first exclude them from the main test run, and then include them in a separate "undefined behaviour" test run? Create a separate test binary?

str4d commented:
If having separate runs is sufficient, then we could by convention append `WithUndefinedBehaviour` to every Google Test that could cause it, modify the `gtest` runner in `./qa/zcash/full_test_suite.py` to call:
```
./src/.zcash-gtest --gtest_filter=-*WithUndefinedBehaviour
```
and then add a new `gtest-ub` runner that calls:

```
./src/.zcash-gtest --gtest_filter=*WithUndefinedBehaviour
```

str4d commented on src/primitives/transaction.h:
```diff
@@ -319,22 +326,30 @@ class CTransaction
+    static_assert(SPROUT_MIN_CURRENT_VERSION >= MIN_TX_VERSION,
+                  "standard rule for tx version should be consistent with network rule");
+
+    static_assert(OVERWINTER_MIN_CURRENT_VERSION >= MIN_TX_VERSION,
```
This is misleading, because once Overwinter activates, `MIN_TX_VERSION` is no longer the minimum version the network rule allows. This was my motivation for having `OVERWINTER_[MIN|MAX]_TX_VERSION` defined in the consensus rules.

bitcartel replied:
I'm adding a constant to consensus.h

daira commented:
@str4d wrote:
> In order for it to not reject such a transaction, both the size of the vin field and its first element would need to exactly collide with the version group ID, and this would only allow a single negative value for the header anyway. So if we do want to try and test this case, we should construct such a malicious transaction as its own serialized format, and then ensure it is rejected; it is otherwise sufficient to check that the parser rejects negative values in the header of an otherwise-Sprout transaction.

I don't see it as being useful to try to test a hypothetical tx format that happens to match the Overwinter one accidentally. We don't do that for Sprout, and I don't think it's a meaningful test.

bitcartel commented:
Pushed a new commit to address most recent review comments.  Filed ticket https://github.com/zcash/zcash/issues/2950 about new gtest runner.

str4d commented on src/main.cpp:
```diff
@@ -840,6 +846,56 @@ unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& in
+    // If Overwinter rules apply:
+    if (isOverwinter) {
+        // Reject transactions with valid version but missing overwinter flag
+        if (tx.nVersion >= CTransaction::OVERWINTER_MIN_CURRENT_VERSION && !tx.fOverwintered) {
```
`tx.nVersion >= OVERWINTER_MIN_TX_VERSION` (now that it is defined)`

str4d commented on src/main.cpp:
```diff
@@ -840,6 +846,56 @@ unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& in
+        }
+
+        // Reject transactions with overwinter flag set but bad version number
+        if (tx.fOverwintered && tx.nVersion < CTransaction::OVERWINTER_MIN_CURRENT_VERSION) {
```
`tx.nVersion < OVERWINTER_MIN_TX_VERSION`

str4d commented on src/main.cpp:
```diff
@@ -866,12 +922,55 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+     *    interprets the sign bit as fOverwintered, so tx.nVersion is always >=0.
+     *    When fOverwintered is set, this function and ContextualCheckTransaction
+     *    will check that tx.nVersion falls witin the range:
+     *        0 <= tx.nVersion < OVERWINTER_MIN_CURRENT_VERSION
```
`OVERWINTER_MIN_TX_VERSION`

str4d commented on src/main.cpp:
```diff
@@ -866,12 +922,55 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+     *    This is unnecessary for Overwinter transactions since the parser now
+     *    interprets the sign bit as fOverwintered, so tx.nVersion is always >=0.
+     *    When fOverwintered is set, this function and ContextualCheckTransaction
+     *    will check that tx.nVersion falls witin the range:
```
`does not fall within the range`

str4d commented on src/main.cpp:
```diff
@@ -866,12 +922,55 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+    // Pre-Overwinter accepts version >=1 transaction. Overwinter only accepts version 3 transactions.
+    // To enable soft-forks of the transaction format, define a range of versions which can be accepted.
+    if (tx.fOverwintered) {
+        if (tx.nVersion < CTransaction::OVERWINTER_MIN_CURRENT_VERSION) {
```
`OVERWINTER_MIN_TX_VERSION`

str4d commented on src/main.cpp:
```diff
@@ -866,12 +928,59 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+            return state.DoS(100, error("CheckTransaction(): overwinter version too low"),
+                REJECT_INVALID, "bad-tx-overwinter-version-too-low");
+        }
+        else if (tx.nVersion < CTransaction::OVERWINTER_MIN_CURRENT_VERSION) {
```
This check shouldn't be in here, as it's a standard rule check which should only be enforced on the mempool, whereas this function is also enforced on the blockchain. If someone duplicated this code to try and create a soft-forking transaction version, this check would eliminate the soft-forkable range between `OVERWINTER_MIN_TX_VERSION` and `OVERWINTER_MIN_CURRENT_VERSION` (which I don't expect would be usable if we follow convention of incrementing the version number, but could conceivably be).

str4d commented on src/main.cpp:
```diff
@@ -866,12 +928,59 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+            return state.DoS(100, error("CheckTransaction(): overwinter version too low for client"),
+                REJECT_INVALID, "bad-tx-overwinter-version-client-too-low");
+        }
+        else if (tx.nVersion > CTransaction::OVERWINTER_MAX_CURRENT_VERSION ) {
```
Define and use an `OVERWINTER_MAX_TX_VERSION` (and don't include `OVERWINTER_MAX_CURRENT_VERSION` here for the same reason as above).

str4d commented on src/main.cpp:
```diff
@@ -866,12 +928,59 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+     *    interprets the sign bit as fOverwintered, so tx.nVersion is always >=0.
+     *    When fOverwintered is set, this function and ContextualCheckTransaction
+     *    will check that tx.nVersion does not fall witin the range:
+     *        0 <= tx.nVersion < OVERWINTER_MIN_TX_VERSION
```
Add `OVERWINTER_MAX_TX_VERSION < tx.nVersion <= INT32_MAX`

str4d commented on src/main.cpp:
```diff
@@ -866,12 +928,59 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+     *        INT32_MIN <= tx.nVersion < 0
+     *    This is unnecessary for Overwinter transactions since the parser now
+     *    interprets the sign bit as fOverwintered, so tx.nVersion is always >=0.
+     *    When fOverwintered is set, this function and ContextualCheckTransaction
```
`When fOverwintered is set, and the Overwinter network upgrade has activated, ...`

str4d commented on src/main.cpp:
```diff
@@ -840,6 +846,62 @@ unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& in
+        }
+
+        // Reject Overwintered transactions with version number considered invalid by network
+        if (tx.fOverwintered && tx.nVersion < OVERWINTER_MIN_TX_VERSION) {
```
This is a duplicate of the rule in `CheckTransactionWithoutProofVerification`, and should not be necessary here, as this function should never be called without `CheckTransaction` being called first. But I'm not strongly opposed to it remaining.

str4d commented on src/main.cpp:
```diff
@@ -840,6 +846,62 @@ unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& in
+        }
+
+        // Reject Overwintered transactions with version number too low to be accepted by this client
+        if (tx.fOverwintered && tx.nVersion < CTransaction::OVERWINTER_MIN_CURRENT_VERSION) {
```
This check shouldn't be in here, for the same reason as I give below in `CheckTransactionWithoutProofVerification`.

str4d commented on src/main.cpp:
```diff
@@ -840,6 +846,62 @@ unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& in
+        }
+
+        // Reject Overwintered transactions with version number too high to be accepted by this client
+        if (tx.fOverwintered && tx.nVersion > CTransaction::OVERWINTER_MAX_CURRENT_VERSION) {
```
Define and use an `OVERWINTER_MAX_TX_VERSION` (and don't include `OVERWINTER_MAX_CURRENT_VERSION` here for the same reason as above).

str4d commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -391,3 +391,339 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+        mtx.nVersion = CTransaction::OVERWINTER_MAX_CURRENT_VERSION + 1;
+        UNSAFE_CTransaction tx(mtx);
+        MockCValidationState state;
+        EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "bad-tx-overwinter-version-client-too-high", false)).Times(1);
```
Will become `bad-tx-overwinter-version-too-high`

str4d commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -391,3 +391,339 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+    mtx.vjoinsplit.resize(0);
+    mtx.fOverwintered = true;
+    mtx.nVersion = 3;
+    mtx.nVersionGroupId = OVERWINTER_VERSION_GROUP_ID;
```
Remove this line.

str4d replied:
Thanks GitHub for matching my old comment to a completely new part of the code!

str4d commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -391,3 +391,339 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+        mtx.nVersion = CTransaction::OVERWINTER_MAX_CURRENT_VERSION + 1;
+        UNSAFE_CTransaction tx(mtx);
+        MockCValidationState state;
+        EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "tx-overwinter-version-client-too-high", false)).Times(1);
```
Will definitely change, and this entire test case may disappear depending on the resolution of my other comments.

bitcartel commented on src/rpcrawtransaction.cpp:
```diff
@@ -435,7 +442,8 @@ UniValue createrawtransaction(const UniValue& params, bool fHelp)
 
-    CMutableTransaction rawTx;
+    CMutableTransaction rawTx = CreateNewContextualCMutableTransaction(
+        Params().GetConsensus(), chainActive.Height() + 1);
```
Lots of scripts out there probably already use `createrawtransaction`, on demand, before sending soon after.   I would guess this is the most popular use-case.  Requiring a default would break existing scripts.

So not needing to pass in another parameter, and defaulting to `chainActive.Height() + 1` seems a reasonable option to keep things running.  

To give users control over this, there could be an optional height field.
`createrawtransaction txs addresses [height]`

That might confuse users though as they might mix it up with expiry height.  Speaking of which... do we need a parameter for that here, or is that in the tx expiry PR?
`createrawtransaction txs addresses [expiryheight] [consensusheight]`

Since this is a low level routine, is this better?
`createrawtransaction txs addresses overwintered`

where `overwintered` is an optional json string, containing non-optional fields:
```
{
"branchid": branchid OR "consensusheight": height
"expiryheight": height
}
```
and if this overwintered json string is not present, we default to using `chainActive.Height` to determine the tx version, and for expiry height add the default value.

str4d replied:
Another thing to note is that upstream introduced new positional arguments:

```
createrawtransaction inputs outputs ( locktime ) ( replaceable )
```

They also introduced the ability to use named arguments.

daira replied:
For 1.0.15, let's use ``chainActive.Height() + 1``. Then for 1.1.0, we can pull in the named arguments PR(s) and add a named ``consensusheight`` (or ``consensusHeight``) argument, with ``chainActive.Height() + 1`` as the default.

bitcartel replied:
Filed https://github.com/zcash/zcash/issues/2952

bitcartel commented:
Pushed commit

- only in `IsStandardTx` do we check OVERWINTER_MIN|MAX_CURRENT_VERSION
- only in `CheckTransactionWithoutProofVerification` do we check OVERWINTER_MIN|MAX_TX_VERSION

str4d commented on src/main.cpp:
```diff
@@ -866,12 +910,55 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+     *        INT32_MIN <= tx.nVersion < 0
+     *    This is unnecessary for Overwinter transactions since the parser now
+     *    interprets the sign bit as fOverwintered, so tx.nVersion is always >=0.
+     *    When fOverwintered is set, and Overwinter network upgrade has activated,
```
`s/and Overwinter network upgrade has activated,//`

bitcartel replied:
Don't understand this comment, since I made the change for this other comment.  https://github.com/zcash/zcash/pull/2925#discussion_r168611377

str4d commented on src/main.cpp:
```diff
@@ -866,12 +910,55 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+     * 2. The previous consensus rule checked for negative values within the range:
+     *        INT32_MIN <= tx.nVersion < 0
+     *    This is unnecessary for Overwinter transactions since the parser now
+     *    interprets the sign bit as fOverwintered, so tx.nVersion is always >=0.
```
Continue sentence with:

```
[...] is always >=0,
and when Overwinter is not active ContextualCheckTransaction rejects
transactions with fOverwintered set.
```

str4d commented:
@zkbot try

ebfull commented:
LGTM modulo comments (some of which are outdated.)

I don't like the habit of doing `if (tx.fOverwintered` over and over, if we miss a `!` things can go wrong. I think that stuff should be in branches. (Right now, many of the `tx.fOverwintered` checks are redundant in their own branches!)

bitcartel commented:
Pushed update.
- Added this phrase https://github.com/zcash/zcash/pull/2925#discussion_r168656088 to comment and tidied up surrounding sentence(s).
- Also addressed @ebfull comment and `tx.fOverwintered` checks in a branch.  See change to `main.cpp` here: https://gist.github.com/bitcartel/4069b3db3b82380312974d67082e7402/revisions?diff=split
- @str4d You can see I removed the condition `tx.nVersion == 3` on the expiry height check because:
  - The version has already been validated earlier in the code path
  - The tx expiry threshold value is specified in the tx expiry zip, it is not linked to or defined by the transaction format version (at least not yet).

bitcartel commented:
@zkbot try

str4d commented on src/main.cpp:
```diff
@@ -866,11 +910,50 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+     *    This is unnecessary for Overwinter transactions since the parser now
+     *    interprets the sign bit as fOverwintered, so tx.nVersion is always >=0,
+     *    and when Overwinter is not active ContextualCheckTransaction rejects
+     *    transactions with fOverwintered set.  When Overwinter has activated and
```
Non-blocking: `s/Overwinter has activated and//`.

str4d commented on src/main.cpp:
```diff
@@ -840,6 +846,62 @@ unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& in
+        }
+
+        // Reject transactions intended for Sprout
+        if (!tx.fOverwintered) {
```
I'm fine with it as-is, because it gives more contextual information as to why the transaction was rejected.

str4d commented:
Any remaining issues can be addressed in follow-up PRs.

@daira please assign a ZIP number, update the top comment, and then r+.

ebfull commented on src/bitcoin-tx.cpp:
```diff
@@ -157,7 +157,7 @@ static void RegisterLoad(const string& strInput)
 {
     int64_t newVersion = atoi64(cmdVal);
-    if (newVersion < CTransaction::MIN_CURRENT_VERSION || newVersion > CTransaction::MAX_CURRENT_VERSION)
+    if (newVersion < CTransaction::SPROUT_MIN_CURRENT_VERSION || newVersion > CTransaction::SPROUT_MAX_CURRENT_VERSION)
```
The issue filed is #2949.

daira commented on src/main.cpp:
```diff
@@ -866,11 +910,50 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+                REJECT_INVALID, "bad-tx-overwinter-version-too-low");
+        }
+        if (tx.nVersion > OVERWINTER_MAX_TX_VERSION ) {
+            return state.DoS(100, error("CheckTransaction(): overwinter version too high for client"),
```
Why "for client"? I don't know what that means.

bitcartel replied:
That message needs to be fixed.

str4d replied:
Looks like it's leftover from a previous version. `s/for client//`

daira commented on src/primitives/transaction.h:
```diff
@@ -304,6 +304,13 @@ class CTxOut
+
+// Overwinter version group id
+static constexpr uint32_t OVERWINTER_VERSION_GROUP_ID = 0x03C48270;
+static_assert(OVERWINTER_VERSION_GROUP_ID != 0, "version group id must be non-zero as specified in ZIP");
```
Update this to ZIP 202.

bitcartel replied:
Done

daira commented on src/test/sighash_tests.cpp:
```diff
@@ -93,10 +94,26 @@ void static RandomScript(CScript &script) {
+    if (tx.fOverwintered) {
+        // Versions outside known ranges throw an exception during parsing
+        auto range =
+            CTransaction::OVERWINTER_MAX_CURRENT_VERSION - CTransaction::OVERWINTER_MIN_CURRENT_VERSION;
```
Off-by-one error; this should be ``CTransaction::OVERWINTER_MAX_CURRENT_VERSION - CTransaction::OVERWINTER_MIN_CURRENT_VERSION + 1``. Fixing this would make the ``if (range > 0)`` redundant.

daira replied:
Filed #2954.

ebfull commented:
I noticed that in `CheckTransaction` we now reject transacations larger than the current max Overwinter version. This happens to be a non-contextual check now, but we will need to increase the max version number in Sapling, which means it will become a contextual check then.

I think this PR should move that check into `ContextualCheckTransaction` to reduce the delta between Overwinter and Sapling upgrade logic.

daira commented on src/main.cpp:
```diff
@@ -866,11 +910,50 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+            return state.DoS(100, error("CheckTransaction(): unknown tx version group id"),
+                    REJECT_INVALID, "bad-tx-version-group-id");
+        }
+        if (tx.nExpiryHeight > TX_EXPIRY_HEIGHT_THRESHOLD) {
```
``LOCKTIME_THRESHOLD = 500000000`` is exclusive, i.e. a locktime is a Unix timestamp if it is *less than* 500000000.
But ``TX_EXPIRY_HEIGHT_THRESHOLD = 500000000`` as implemented here (and as specified in [ZIP 203](https://github.com/zcash/zips/pull/131/files)) is inclusive, i.e. the value 500000000 is accepted and treated as a block height rather than a Unix timestamp. This is inconsistent (and would result in a very ugly wart if we decided to extend ``nExpiryHeight`` to allow the timestamp meaning in future).

bitcartel replied:
Good spot, the bitcoin wiki says here https://en.bitcoin.it/wiki/Protocol_documentation#tx
```
< 500000000 | Block number at which this transaction is unlocked
>= 500000000 | UNIX timestamp at which this transaction is unlocked
```
Will update the code.  @arcalinea can you please do the same with the ZIP?  Thanks.

bitcartel commented:
Pushed update based on review comments.  Changes are here: https://gist.github.com/bitcartel/1072f276691024998f384c14e700ced7

str4d commented:
Timing out on re-ACK-tion from @daira and @ebfull - the update addresses both their latest comments.

Thanks all!

@zkbot r+

ebfull commented:
re-ACK.

daira commented on src/test/sighash_tests.cpp:
```diff
@@ -234,7 +253,16 @@ BOOST_AUTO_TEST_CASE(sighash_from_data)
           CValidationState state;
-          if (tx.nVersion < MIN_TX_VERSION) {
+          if (tx.fOverwintered) {
+              if (tx.nVersion == 3 && tx.nExpiryHeight > TX_EXPIRY_HEIGHT_THRESHOLD) {
```
s/>/>=/

daira replied:
Also add a test to make sure that the ``== TX_EXPIRY_HEIGHT_THRESHOLD`` case is exercised. (It isn't currently or this PR wouldn't have merged!)

daira replied:
Filed #2959 to make sure this doesn't get lost.
